src/java/org/apache/cassandra/io/sstable/format/big/BigTableWriterBuilder.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.io.sstable.format.big;
+
+import org.apache.cassandra.db.lifecycle.LifecycleNewTracker;
+import org.apache.cassandra.io.sstable.DataComponent;
+import org.apache.cassandra.io.sstable.Descriptor;
+import org.apache.cassandra.io.sstable.format.SortedTableWriterBuilder;
+import org.apache.cassandra.io.util.SequentialWriter;
+import org.apache.cassandra.utils.Throwables;
+
+public class BigTableWriterBuilder extends SortedTableWriterBuilder<RowIndexEntry, BigFormatPartitionWriter, BigTableWriter, BigTableWriterBuilder>
+{
+    private RowIndexEntry.IndexSerializer rowIndexEntrySerializer;
+    private BigTableWriter.IndexWriter indexWriter;
+    private SequentialWriter dataWriter;
+    private BigFormatPartitionWriter partitionWriter;
+
+    public BigTableWriterBuilder(Descriptor descriptor)
+    {
+        super(descriptor);
+    }
+
+    @Override
+    public SequentialWriter getDataWriter()
+    {
+        return dataWriter;
+    }
+
+    @Override
+    public BigFormatPartitionWriter getPartitionWriter()
+    {
+        return partitionWriter;
+    }
+
+    public RowIndexEntry.IndexSerializer getRowIndexEntrySerializer()
+    {
+        return rowIndexEntrySerializer;
+    }
+
+    public BigTableWriter.IndexWriter getIndexWriter()
+    {
+        return indexWriter;
+    }
+
+    @Override
+    protected BigTableWriter buildInternal(LifecycleNewTracker lifecycleNewTracker)
+    {
+        try
+        {
+            rowIndexEntrySerializer = new RowIndexEntry.Serializer(descriptor.version, getSerializationHeader());
+            dataWriter = DataComponent.buildWriter(descriptor,
```

Comment:
I did it that way in order to avoid long arguments lists on constructor. Also having say, data file handle and partition writer passed as separate fields in the constructor would theoretically let pass a handle to a different file than the provided partition writer operates on. As far as I understand, those two parameters are inherently related and passing them always together (in a single builder) is more consistent to me.

The resource fields are cleaned in finally block, so the builder can be safely reused.

Reply:
Well, usually builders are so that end users don't use long argument lists; the internals can, and in this case it makes better sense that they do. If necessary, the builder can be made a static inner class so that the relevant constructor is private.

Reply:
We cannot do a private constructor with this approach because we have a hierarchy - some components are consumed by the constructor of `SortedTableWriter` and the others by `BigTableWriter`. Obviously such a constructor in `SortedTableWriter` would have to be available for its descendants. 

I understand your point, though. I know it is not a perfect approach but I also don't like having different approaches for writer and reader. I think I can try to meet your worries by doing it a bit differently - while not retiring passing stuff in a builder, I can refactor the initialization a bit: f932e370396a287c4293c6698dd9590d5258b72b

Reply:
This is overdoing it -- making sure that `buildInternal` cannot be called more than once would suffice.